### ISSUE TITLE
clusters is not iterable

### ISSUE THREAD

**Issue by justathoughtor2** [paiqo/Databricks-VSCode#8]
The clusters tab is completely blank. When I attempt to refresh it, a notification appears in VS Code stating "clusters is not iterable." The only output from the extension appears to be success messages:

Logger initialized!
Validating settings ...
Settings validated!
Initializing Databricks API Service ...
Databricks API Service initialized!

At this point, since I can't connect to any cluster, this extension is essentially unusable. Any assistance would be greatly appreciated. According to VS Code, I'm using version 0.3.1 of the extension.

Thanks!
Jesse

**Comment by gbrueckl** [comment]
Hi Jesse,

do the other tabs like "Workspace", "DBFS", work correctly?
I assume you do have clusters defined in your DAtabricks workspace - right?

**Comment by FlorianPydde** [comment]
Hi all,

I am experiencing the same issue as Jesse. Essentially, there is neither clusters nor the workspace showing up in the tabs. I have multiple folders and clusters in the workspace. 

Any help is appreciated.
Florian

**Comment by gbrueckl** [comment]
On which OS are you? Windows? Linux? MaxOS?
it is working fine on my Windows machine
also - Azure or AWS?

**Comment by FlorianPydde** [comment]
I am on Windows using Azure

**Comment by gbrueckl** [comment]
i created a new version with better logging - it is not in the VSCode gallery yet but you can download and test it from here:
https://github.com/paiqo/Databricks-VSCode/releases/tag/v0.3.3

**Comment by justathoughtor2** [comment]
I'm on Windows with AWS. The system can connect to the root API successfully, and there are both clusters and workspaces configured in Databricks. But they don't show up in the extension yet. I'll try the new version and see what gets logged.

**Comment by justathoughtor2** [comment]
I got it working! The more verbose logging helped nail down the problem. There was a 403 being thrown. Apparently my token had expired (unbeknownst to me). Creating a new token and restarting VS Code fixed the problem. Thanks for the help!

**Comment by gbrueckl** [comment]
great, good to hear - how about you @FlorianPydde ?

**Comment by cgrisham** [comment]
On Windows trying to connect to Azure Databricks
I'm having the same issue. But I have no items in workspaces or clusters or any window. 

Logger initialized!
Validating settings ...
Settings validated!
Initializing Databricks API Service ...
Databricks API Service initialized!
GET 2.0/workspace/list
Params:{"params":{"path":"/"}}
GET 2.0/clusters/list
Params:null
GET 2.0/jobs/list
Params:null
GET 2.0/dbfs/list
Params:{"params":{"path":"/"}}
GET 2.0/secrets/scopes/list
Params:null


is what I get from the logger, and then a 6 strings of HTML which appears to be the sign in screen.

**Comment by gbrueckl** [comment]
@cgrisham I added better error handling in v0.3.4
https://github.com/paiqo/Databricks-VSCode/releases/tag/v0.3.4

can you test it on your local machine and post the output

**Comment by cgrisham** [comment]
Besides the log output I provided above the only other errors I know to provide are these. If you need something else let me know. 

```
notificationsAlerts.ts:38 TypeError: clusters is not iterable
	at DatabricksClusterTreeItem.<anonymous> (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.3\out\databricksApi\clusters\DatabricksClusterTreeItem.js:99:37)
	at Generator.next (<anonymous>)
	at fulfilled (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.3\out\databricksApi\clusters\DatabricksClusterTreeItem.js:5:58)
	at processTicksAndRejections (internal/process/task_queues.js:85:5)
```

```
notificationsAlerts.ts:38 TypeError: Cannot read property 'sort' of undefined
	at Function.sortArrayByProperty (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.3\out\helpers\Helper.js:83:23)
	at Function.<anonymous> (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.3\out\databricksApi\databricksApiService.js:87:29)
	at Generator.next (<anonymous>)
	at fulfilled (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.3\out\databricksApi\databricksApiService.js:5:58)
	at processTicksAndRejections (internal/process/task_queues.js:85:5)
```

**Comment by cgrisham** [comment]
[7-paiqo.databricks-vscode.log](https://github.com/paiqo/Databricks-VSCode/files/4416092/7-paiqo.databricks-vscode.log)

Is what the extension log shows.

Console error is reporting this if it's helpful

```
TypeError: clusters is not iterable
	at DatabricksClusterTreeItem.<anonymous> (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.4\out\databricksApi\clusters\DatabricksClusterTreeItem.js:99:37)
	at Generator.next (<anonymous>)
	at fulfilled (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.4\out\databricksApi\clusters\DatabricksClusterTreeItem.js:5:58)
	at processTicksAndRejections (internal/process/task_queues.js:85:5)
```

```
TypeError: Cannot read property 'sort' of undefined
	at Function.sortArrayByProperty (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.4\out\helpers\Helper.js:83:23)
	at Function.<anonymous> (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.4\out\databricksApi\databricksApiService.js:118:29)
	at Generator.next (<anonymous>)
	at fulfilled (c:\Users\cgrisham\.vscode\extensions\paiqo.databricks-vscode-0.3.4\out\databricksApi\databricksApiService.js:5:58)
	at processTicksAndRejections (internal/process/task_queues.js:85:5)
```

**Comment by gbrueckl** [comment]
that very much looks like a missconfiguration
can you check the setting API Root Url setting and if it is something like https://westeurope.azuredatabricks.net

**Comment by cgrisham** [comment]
That was it I had _https://eastus.azuredtabricks.net_**_/_**  as the value. Once I removed the ending **/** it worked fine.

**Comment by gbrueckl** [comment]
ok, good catch!
will also fix the code so it works with or without trailing slash!

**Comment by gbrueckl** [comment]
also the API root URL issue was fixed now with v0.3.5 which was also released to the gallery yesterday

if there are still issues with the new version please open a new ticket
